feat: add option to change the event date/time format

laurelfulford opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds an option to change the four front-facing date formats used by the Events post type:

* The full event date
* The full event time
* Two partial event dates, to reduce duplication (one without the year, and one without the month)

This change improves the ability to translate the event date, on top of #391, since date and time formats often need to change to properly translate them. 

See 1200550061930446-as-1204498431816519

### How to test the changes in this Pull Request:

1. Apply the PR.
2. Create a few event posts. 
3. Navigate to WP Admin > Listings > Settings; confirm there is now a section that allows you to change the event date format, and that it uses the plugin's original defaults:

![image](https://github.com/Automattic/newspack-listings/assets/177561/451bd5bc-7208-47a1-a72c-460249f87100)

4. Try changing the date and time formats; confirm that they're picked up by single Events and Events displayed in the Curated Listings block. 
5. Create an event with a hidden time and day range; confirm that the partial dates you set show up (eg. August 21 - 28, 2023).

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

leogermani reviewed (changes requested):
It works well, but it does not pick the general settings by default. The defaults are still hardcoded.

I think we need a function to dynamically get the value from the plugin options, if present, and fallback to the systems option.

Do you want me to take it over?

laurelfulford commented:
Thanks for taking a look @leogermani!

> It works well, but it does not pick the general settings by default. The defaults are still hardcoded.

This was deliberate, but maybe shouldn't be: we have a small number of publishers (< 5) who include the time in the date format as a work around to display both, so I was trying to avoid that being pulled into the Events. But since it can be overridden, it's probably an unnecessary caution vs. the experience of having the date format "just work" in all other cases. 

> Do you want me to take it over?

If you have the bandwidth over the next couple days, sure -- thank you! Otherwise I can circle back early next week 🙂

laurelfulford commented:
I've updated the PR to pull the date format from Settings > General as the default. 

Then I tried to get clever with the formats without the Year or Month and use Regex (and some ChatGPT!) to update the formats 😬 This seems to work in my testing, but I'm not sure if it's absolute overkill, especially for this relatively small feature (and if it even absolutely would work in all scenerios!). 

@dkoo When you have a moment, can I get your thoughts on the above? I think Leo's suggestion is a good idea but I'm not sure it works great with the truncated date formats the plugin currently uses, and if it would make sense to try to update them like this, leave them as the original defaults, or just let the plugin handle it's own dates (the downside there is that it's an extra step to get the format updated, which isn't a great user experience -- it's much nicer if it "just works" and picks up what the site is otherwise using). 

The regex should at least correctly update the default date options offered by WordPress, but isn't going to be able to figure out any format: 

* `F j, Y` -- should update to `F j` when the year is removed, or `j, Y` when the month is removed.
* `Y-m-d` -- should update to `m-d` when the year is removed, or `Y-d` when the month is removed.
* `m/d/Y` -- should update to `m/d` when the year is removed, or `d/Y` when the month is removed.
* `d/m/Y` -- should update to `d/m` when the year is removed, or `d/Y` when the month is removed. 

The possible downsides include that this won't work with all possible date formats, and the resulting formats may not make sense even though they make sense with the original `F j, Y` format -- another option could be to only use the truncated no-month and no-year date when that format's being used.

dkoo commented:
Hmm, this is still not working in all cases, and I blame myself for trying to get fancy with showing human-readable date ranges while not accounting for non-US date formatting. For example, set your date format to `Y-m-d` and then set a start + end date on consecutive days within the same month—the date will look off. Let me try to dig into this further to see if we can preserve the human readability with all of these different options. Worst-case, maybe we don't try to parse the dates in a human-readable way unless the date format is the default `F j, Y`.

laurelfulford commented:
@dkoo That all makes sense 🤦‍♀️ I didn't even get into testing what the different formats would look like paired with the human-readable logic, that's a mess, thanks for catching that! 

Working on updating this now!

laurelfulford commented:
@dkoo Ok, I think this is ready to roll:

* I removed the extra human-readable date fields from the Listing Settings
* I added a check to the code to use the human-readable range only when the selected date format was `F j, Y`
* I found and fixed an error I made that caused the end date to pull today's date 😐 In my testing the dates seem to work now, but if you have a chance to look at that to make sure I haven't done anything silly, it'd be appreciated!

dkoo reviewed (approved):
This is working well for me now! Thanks for the updates, @laurelfulford

laurelfulford commented:
Awesome - thanks for mergin', @dkoo! 🥳
